### PR TITLE
Move metaclass initialization to python3

### LINKED ISSUE #7932: Usage of Python 2 `__metaclass__`
There are 7 instances of `__metaclass__` in Synapse, which does nothing in Python 3.
These should probably be changed to use `class Foo(metaclass=abc.ABCMeta)`.

```
synapse/handlers/room_member.py
48:    __metaclass__ = abc.ABCMeta

synapse/server.py
141:    __metaclass__ = abc.ABCMeta

synapse/replication/http/_base.py
77:    __metaclass__ = abc.ABCMeta

synapse/storage/data_stores/main/account_data.py
41:    __metaclass__ = abc.ABCMeta

synapse/storage/data_stores/main/receipts.py
41:    __metaclass__ = abc.ABCMeta

synapse/storage/data_stores/main/push_rule.py
80:    __metaclass__ = abc.ABCMeta

synapse/storage/data_stores/main/stream.py
251:    __metaclass__ = abc.ABCMeta
```

### PR DESCRIPTION
**Pull request by ShadowJonathan**
### Pull Request Checklist

This moves every instance of `__metaclass__ = Y` to be `class X(metaclass=Y)`-based

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Fixes #7932

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: changelog.d/8326.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Moves every instance of `__metaclass__ = Y` to be `class X(metaclass=Y)`-based.
```
Comment (clokep) [inline changelog.d/8326.misc:1]: Can you update this to be a bit more descriptive, maybe:

> Update the use of `metaclass` to be compatible with Python 3.

Not sure if that really makes sense or not though.

Also feel free to add "Contributed by <your name>." At the end if you'd like!

Comment (ShadowJonathan) [reply changelog.d/8326.misc:1]: Ah, right, I wanted to be a tad more technical in my wording, but this makes more sense, thanks!

Comment (clokep) [reply changelog.d/8326.misc]: It's tough, usually the people reading these are not developers (not people who necessarily know Python well), so we try not to make them too deep in the weeds.

Then again, `misc` changes tend to do that. 👍

**Review by clokep — APPROVED** [review]
Looks OK if the build passes. 👍 Thanks!

**Pull request merged**